modules/xphoto/src/inpainting.cpp

```diff
@@ -66,6 +77,820 @@ namespace cv
+        int height = bgr.rows;
+        int width = bgr.cols;
+        Y = cv::Mat::zeros(height, width, CV_8U);
+        Cb = cv::Mat::zeros(height, width, CV_8U);
+        Cr = cv::Mat::zeros(height, width, CV_8U);
+
+        cv::Mat channels_bgr[3];
+        cv::split(bgr, channels_bgr);
+        uchar* rData = (uchar*)channels_bgr[2].data;
+        uchar* gData = (uchar*)channels_bgr[1].data;
+        uchar* bData = (uchar*)channels_bgr[0].data;
+        uchar* yData = (uchar*)Y.data;
+        uchar* cbData = (uchar*)Cb.data;
+        uchar* crData = (uchar*)Cr.data;
+        for (int y = 0; y < height; ++y)
+        {
+            for (int x = 0; x < width; ++x)
+            {
+                uchar R = rData[y*channels_bgr[2].step1() + x];
+                uchar G = gData[y*channels_bgr[1].step1() + x];
+                uchar B = bData[y*channels_bgr[0].step1() + x];
+                yData[y*Y.step1() + x] = cv::saturate_cast<uchar>(0.256788235294118 * R + 0.504129411764706 * G + 0.0979058823529412 * B + 16);
+                cbData[y*Cb.step1() + x] = cv::saturate_cast<uchar>(-0.148223529411765 * R - 0.290992156862745* G + 0.439215686274510 * B + 128.0);
+                crData[y*Cr.step1() + x] = cv::saturate_cast<uchar>(0.439215686274510 * R - 0.367788235294118 * G - 0.0714274509803922 * B + 128);
+            }
+        }
+    }
+
+
+    static void
+    icvYCbCr2BGR(cv::Mat& Y, cv::Mat& Cb, cv::Mat& Cr, cv::Mat& bgr)
+    {
+        // same behavior as matlab ycbcr2rgb, when ycbcr image is of type uint8
+        int height = Y.rows;
+        int width = Y.cols;
+
+        uchar* bgrData = (uchar*)bgr.data;
+        double* yData = (double*)Y.data;
+        double* cbData = (double*)Cb.data;
+        double* crData = (double*)Cr.data;
+        for (int y = 0; y < height; ++y)
+        {
+            for (int x = 0; x < width; ++x)
+            {
+                double Y_curr = yData[y*Y.step1() + x];
+                double Cb_curr = cbData[y*Cb.step1() + x];
+                double Cr_curr = crData[y*Cr.step1() + x];
+                bgrData[y*bgr.step1() + 3 * x] = cv::saturate_cast<uchar>(1.16438356164384 * Y_curr + 2.01723263955646* Cb_curr + 0.000003054 * Cr_curr - 276.836305795032);
+                bgrData[y*bgr.step1() + 3 * x + 1] = cv::saturate_cast<uchar>(1.16438356164384 * Y_curr - 0.391762539941450 * Cb_curr - 0.812968292162205 * Cr_curr + 135.575409522967);
+                bgrData[y*bgr.step1() + 3 * x + 2] = cv::saturate_cast<uchar>(1.16438356164384 * Y_curr + 3.01124397411008e-07 * Cb_curr + 1.59602688733570 * Cr_curr - 222.921617109194);
+            }
+        }
+    }
+
+
+    static void
+    icvSgnMat(const cv::Mat& src, cv::Mat& dst) {
+        dst = cv::Mat::zeros(src.size(), CV_64F);
+        double* srcData = (double*)src.data;
+        for (int y = 0; y < src.rows; ++y)
+        {
+            for (int x = 0; x < src.cols; ++x)
+            {
+                double curr_val = srcData[y*src.step1() + x];
```

Comment:
Please avoid manual calculation through steps:
```
CV_CheckTypeEQ(CV_64F, src.type(), "");
for (y...)
{
    double curr_row = src.ptr<double>(y);
    for (x...)
    {
        double curr_val = curr_row[x];
        ...
    }
}
```

Reply:
I used ``src.at<double>(y,x)`` instead of the manual calculation. I hope that is fine as well.